line-height should use LengthType::Normal to represent 'normal'.

AlexsanderDamaceno opened the pull request:
#### 1f9a7eaf0b2e916b9b2ad6edd4955c35f8b3cebe
<pre>
line-height should use LengthType::Normal to represent &apos;normal&apos;.
<a href="https://bugs.webkit.org/show_bug.cgi?id=277249">https://bugs.webkit.org/show_bug.cgi?id=277249</a>

Reviewed by Tim Nguyen.

The line height computation uses a negative value to represent &apos;normal&apos; in the RenderStyle.
Replace negative value check with isNormal() to check if line height is &apos;normal&apos; as LengthType::Normal was introduced.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::computeLineHeight const):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::optionalLineHeight):
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
(WebCore::Layout::InlineLevelBox::isPreferredLineHeightFontMetricsBased const):
* Source/WebCore/layout/formattingContexts/inline/InlineQuirks.cpp:
(WebCore::Layout::InlineQuirks::initialLetterAlignmentOffset const):
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialSpecifiedLineHeight):
(WebCore::RenderStyle::initialLineHeight):
* Source/WebCore/rendering/TextAutoSizing.cpp:
(WebCore::TextAutoSizingValue::adjustTextNodeSizes):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueLineHeight):
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::adjustSelectListButtonStyle):

Canonical link: <a href="https://commits.webkit.org/282043@main">https://commits.webkit.org/282043@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0ca09004255aa3242351d45bc042789b5672208e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61850 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41204 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14442 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65829 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12395 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63969 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48890 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12666 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49841 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8580 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64919 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38263 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53566 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30673 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34907 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10799 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11326 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56716 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11103 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67558 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5793 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10835 "Found 1 new test failure: workers/worker-to-worker.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57223 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5818 "Built successfully") | [⏳ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/macOS-Release-WK2-Stress-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57460 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13759 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4749 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37004 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38088 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39184 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37833 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -812,7 +812,7 @@ inline bool RenderStyle::useTouchOverflowScrolling() const { return m_rareInheri
 
 #if ENABLE(TEXT_AUTOSIZING)
-inline Length RenderStyle::initialSpecifiedLineHeight() { return { -100.0f, LengthType::Percent }; }
+inline Length RenderStyle::initialSpecifiedLineHeight() { return { -100.0f, LengthType::Normal }; }
```
```suggestion
inline Length RenderStyle::initialSpecifiedLineHeight() { return LengthType::Normal; }
```

This should just work.

nt1m commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -868,7 +868,7 @@ inline bool RenderStyle::hasInlineColumnAxis() const
 {
     // Returning -100% here means the line-height is not set.
-    return { -100.0f, LengthType::Percent };
+    return { -100.0f, LengthType::Normal };
```
Ditto

nt1m commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -868,7 +868,7 @@ inline bool RenderStyle::hasInlineColumnAxis() const
 inline Length RenderStyle::initialLineHeight()
 {
     // Returning -100% here means the line-height is not set.
```
I would drop this comment

nt1m commented on Source/WebCore/rendering/TextAutoSizing.cpp:
```diff
@@ -160,7 +160,7 @@ auto TextAutoSizingValue::adjustTextNodeSizes() -> StillHasNodes
         auto newParentStyle = cloneRenderStyleWithState(parentStyle);
-        newParentStyle.setLineHeight(lineHeightLength.isNegative() ? Length(lineHeightLength) : Length(lineHeight, LengthType::Fixed));
+        newParentStyle.setLineHeight(lineHeightLength.isNormal() ? Length(lineHeightLength) : Length(lineHeight, LengthType::Fixed));
         newParentStyle.setSpecifiedLineHeight(Length { lineHeightLength });
```
I think this line needs to change to:
```
newParentStyle.setSpecifiedLineHeight(lineHeightLength.isNormal() ? Length { LengthType::Normal } : Length { lineHeightLength });
```

Otherwise, it discards the "normal" type. The old code used to work, because the other checks just looked at whether the value was negative.

nt1m replied:
That hopefully should fix all the iOS failures.

AlexsanderDamaceno replied:
@nt1m, I did not understood why it discards "normal" type, as  Length(lineHeightLength) copy constructor  sets  type to LengthType::Normal.

nt1m replied:
Hmm you're right, let me take a deeper look

nt1m reviewed (approved):
r=me if all the checks are green

AlexsanderDamaceno committed:
line-height should use LengthType::Normal to represent 'normal'.
https://bugs.webkit.org/show_bug.cgi?id=277249

Reviewed by Tim Nguyen.

The line height computation uses a negative value to represent 'normal' in the RenderStyle.
Replace negative value check with isNormal() to check if line height is 'normal' as LengthType::Normal was introduced.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::computeLineHeight const):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::optionalLineHeight):
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
(WebCore::Layout::InlineLevelBox::isPreferredLineHeightFontMetricsBased const):
* Source/WebCore/layout/formattingContexts/inline/InlineQuirks.cpp:
(WebCore::Layout::InlineQuirks::initialLetterAlignmentOffset const):
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialSpecifiedLineHeight):
(WebCore::RenderStyle::initialLineHeight):
* Source/WebCore/rendering/TextAutoSizing.cpp:
(WebCore::TextAutoSizingValue::adjustTextNodeSizes):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueLineHeight):
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::adjustSelectListButtonStyle):

Canonical link: https://commits.webkit.org/282043@main
